contracts/eosio.system/src/wps.cpp

```diff
@@ -0,0 +1,774 @@
+        if( voter->is_proxy ) {
+            check(false, "Proxies can't vote for worker proposals");
+        }
+
+        std::map<name, std::pair<double, bool /*new*/> > proposal_deltas;
+        if ( voter->last_vote_weight > 0 ) {
+            if(wpsvoter != _wpsvoters.end()){
+                for( const auto& p : wpsvoter->proposals ) {
+                    auto& d = proposal_deltas[p];
+                    d.first -= voter->last_vote_weight;
+                    d.second = false;
+                }
+            }
+        }
+
+        if( new_vote_weight >= 0 ) {
+            for( const auto& p : proposals ) {
+                auto& d = proposal_deltas[p];
+                d.first += new_vote_weight;
+                d.second = true;
+            }
+        }
+
+        const auto ct = current_time_point();
+        for( const auto& pd : proposal_deltas ) {
+            auto pitr = _proposals.find( pd.first.value );
+            if( pitr != _proposals.end() ) {
+
+                check( (*pitr).status != PROPOSAL_STATUS::PENDING, "The proposal is not currently on vote" );
+
+                time_point_sec current_time = current_time_point();
+                wps_env_singleton _wps_env(get_self(), get_self().value);
+                auto wps_env = _wps_env.get();
+                uint32_t duration_of_voting = wps_env.duration_of_voting * seconds_per_day;
+
+                if(time_point_sec(time_point(current_time - (*pitr).vote_start_time)) >= time_point_sec(duration_of_voting)) {
+                    _proposals.modify(pitr, same_payer, [&](auto &proposal) {
+                        proposal.status = PROPOSAL_STATUS::REJECTED;
+                    });
+                }
+                else{
+                    _proposals.modify( pitr, same_payer, [&]( auto& p ) {
+                        p.total_votes += pd.second.first;
+                        if ( p.total_votes < 0 ) { // floating point arithmetics can give small negative numbers
+                            p.total_votes = 0;
+                        }
+                    });
+
+                    auto total_activated_vote = stake2vote2(_gstate.total_activated_stake);
+                    if((*pitr).total_votes > (double) (total_activated_vote / ((double) (100/wps_env.total_voting_percent)))){
+                        if((*pitr).status == PROPOSAL_STATUS::ON_VOTE){
+                            _proposals.modify( pitr, same_payer, [&]( auto& p ) {
+                                p.status = PROPOSAL_STATUS::FINISHED_VOTING;
+                            });
+                        }
+                    }
+                }
+            } else {
+                if( pd.second.second ) {
+                    check( false, ( pd.first.to_string() + "does not have an existing proposal" ).data() );
+                }
+            }
+        }
+
+        _voters.modify( voter, same_payer, [&]( auto& av ) {
+            av.last_vote_weight = new_vote_weight;
```

Comment:
making this a prop in the wpsvoters table, removing this block and reflecting it in the if-else below will probably solve the interactions

Reply:
Address in new commits